nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DetectDuplicateRecord.java

```diff
@@ -0,0 +1,620 @@
+                CacheValue cacheValue = cache.get(cacheKey, keySerializer, cacheValueDeserializer);
+                Long durationMS = context.getProperty(AGE_OFF_DURATION).asTimePeriod(TimeUnit.MILLISECONDS);
+
+                if(durationMS != null && (now >= cacheValue.getEntryTimeMS() + durationMS)) {
+                    boolean status = cache.remove(cacheKey, keySerializer);
+                    logger.debug("Removal of expired cached entry with key {} returned {}", new Object[]{cacheKey, status});
+                } else {
+                    serializableFilter = cacheValue.getFilter();
+                }
+            }
+
+            final FilterWrapper filter = FilterWrapper.create(serializableFilter);
+
+            final String recordHashingAlgorithm = context.getProperty(RECORD_HASHING_ALGORITHM).getValue();
+            final MessageDigest messageDigest = DigestUtils.getDigest(recordHashingAlgorithm);
+            final Boolean matchWholeRecord = context.getProperties().keySet().stream().noneMatch(p -> p.isDynamic());
+
+            final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+            final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+            final RecordReader reader = readerFactory.createRecordReader(flowFile.getAttributes(), session.read(flowFile), logger);
+
+            FlowFile nonDuplicatesFlowFile = session.create(flowFile);
+            final FlowFile duplicatesFlowFile = session.create(flowFile);
+            final RecordSchema writeSchema = writerFactory.getSchema(flowFile.getAttributes(), reader.getSchema());
+            final RecordSetWriter nonDuplicatesWriter = writerFactory.createWriter(getLogger(), writeSchema, session.write(nonDuplicatesFlowFile));
+            final RecordSetWriter duplicatesWriter = writerFactory.createWriter(getLogger(), writeSchema, session.write(duplicatesFlowFile));
+
+            nonDuplicatesWriter.beginRecordSet();
+            duplicatesWriter.beginRecordSet();
+            Record record;
+
+            while ((record = reader.nextRecord()) != null) {
+                String recordValue;
+
+                if (matchWholeRecord) {
+                    recordValue = record.getSerializedForm().get().getSerialized().toString();
+                } else {
+                    final List<String> fieldValues = new ArrayList<>();
+                    for (final String recordPathText : recordPaths) {
+                        final PropertyValue recordPathPropertyValue = context.getProperty(recordPathText);
+                        final RecordPath recordPath = recordPathCache.getCompiled(recordPathText);
+                        final RecordPathResult result = recordPath.evaluate(record);
+                        fieldValues.addAll(result.getSelectedFields()
```

Comment:
@joewitt @markap14 @bbende  does the record API have any guarantees that the order would be consistent through multiple runs of the same data like this?

Reply:
If not, I can always order the resolved RecordPath values first before hashing.

Reply:
Addressed this in latest commit.